### PR TITLE
GoldPC implementation (and DynamicNPC Create Request)

### PR DESCRIPTION
**Pull request by guilherme-gm**
### Pull Request Prelude
- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
This PR focuses on implementing GoldPC UI (aka Mileage). I don't have too much information about it, so I tried to make it customizable enough so it may be used as people seem fit (or adjusted if needed). I chose to make it account-based, but it should not be too hard to make it character-based.

Given that, this UI also uses the "DynamicNPC Create Request" packets, which was not implemented in Herc.

The changes were split into a few self-contained commits, so it may be easier to review it by commit instead of all files at once.

**1. Adds the DynamicNPC Create Request packets/handling**
I don't know whether this is used for something else too, so I made it as isolated as possible from GoldPC. The pair of packets simply consists of a named "request" and an answer. I followed how private airship and inventory expansion works and chose to implement it by calling a specific NPC that handles the name and provides an answer.

For this first commit, it will simply log that it found nothing (as there is nothing using it yet).

**2. Adds a basic implementation of GoldPC**
This basic implementation covers the core functionality, with a single mode.

**3. Adds ways to customize GoldPC**
This includes a database for modes and script command to dynamically change it.

**4. GoldPC NPC**
**This part is the totally custom** (I am ok if you prefer to not include it). I have never seem this system/NPC so the only information I had is what is described in the launch note, and I made the NPC from that.

It includes a NPC that is spawned for the player upon clicking the GoldPC button (using DynamicNPC Create Request). This NPC will tell the player how many points they have and allow changing them for some rewards.

Hercules doesn't have the package items they use and is also missing most of the package contents, so in order to keep this PR as small as possible, I decided to put some basic dummy items for very high prices. I will see if I can make a PR adding those items later.

**5. Docs**
A doc file explaining how it works and how to customize it. I guess it will help someone that is not checking the PR get around the functionalities.

----

The PR ended up quite long, so if there is something I can do to make it easier for review, let me know.

Special thanks to @4144 for helping with dates :)

**Issues addressed:** None, I think

**References:**
- [kRO update note](https://web.archive.org/web/20140625231844/http://ro.gnjoy.com/news/update/View.asp?seq=153&curpage=1#popup1) (web archive link since the page is gone)
- [kRO patch notes of GoldPC pre-release (2014-06-11)](https://ro.gnjoy.com/news/notice/View.asp?BBSMode=10001&seq=6419&curpage=39)
- [kRO patch notes of GoldPC release (2014-06-18)](https://ro.gnjoy.com/news/notice/View.asp?BBSMode=10001&seq=6422&curpage=39)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/packets_struct.h [hunk 1]
```diff
@@ -5817,6 +5817,22 @@ struct PACKET_ZC_SPECIALPOPUP {
...
 
+#if PACKETVER >= 20140430
+struct PACKET_CZ_DYNAMICNPC_CREATE_REQUEST {
+	uint16 PacketType;
+	char name[NAME_LENGTH];
+} __attribute__((packed));
+DEFINE_PACKET_HEADER(CZ_DYNAMICNPC_CREATE_REQUEST, 0x0a16);
+#endif // PACKETVER >= 20140430
+
+#if PACKETVER >= 20140430
```
Comment (4144) [inline src/map/packets_struct.h:5840]: this packet supported from 20140611

File: src/map/clif.c [hunk 2]
```diff
@@ -25700,6 +25700,43 @@ static void clif_special_popup(struct map_session_data *sd, int popupId)
...
+		ShowError("%s: event '%s' not found, operation failed.\n", __func__, evname);
+	}
+#else
+	ShowWarning("%s: Dynamic NPC Create is not supported in this client version, possible packet manipulation.\n", __func__);
+#endif // 20140430
+}
+
+static void clif_dynamicnpc_create_result(struct map_session_data *sd, enum dynamicnpc_create_result result)
+{
+#if PACKETVER >= 20140430
```
Comment (4144) [inline src/map/clif.c:25731]: this packet supported from 20140611

File: src/map/clif.c [hunk 3]
```diff
@@ -25700,6 +25700,43 @@ static void clif_special_popup(struct map_session_data *sd, int popupId)
...
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DYNAMICNPC_CREATE_RESULT));
+	struct PACKET_ZC_DYNAMICNPC_CREATE_RESULT *p = WFIFOP(fd, 0);
+	p->PacketType = HEADER_ZC_DYNAMICNPC_CREATE_RESULT;
+	p->result = result;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DYNAMICNPC_CREATE_RESULT));
+#endif // 20140430
```
Comment (4144) [inline src/map/clif.c:25740]: this packet supported from 20140611

File: src/map/clif.h [hunk 4]
```diff
@@ -907,6 +907,13 @@ enum enchantui_status {
 	ENCHANTUI_FAILURE,
 };
 
+enum dynamicnpc_create_result {
+	DYNAMICNPC_RESULT_SUCCESS = 0,
+	DYNAMICNPC_RESULT_UNKNOWN = 1,
+	DYNAMICNPC_RESULT_NOT_FOUND = 2,
```
Comment (4144) [inline src/map/clif.h:913]: in official enum look like used ``NOT_EXIST``

File: src/map/packets_struct.h [hunk 5]
```diff
@@ -5817,6 +5817,18 @@ struct PACKET_ZC_SPECIALPOPUP {
 DEFINE_PACKET_HEADER(ZC_SPECIALPOPUP, 0x0bbe);
 #endif  // PACKETVER_MAIN_NUM >= 20221005
 
+#if PACKETVER >= 20140430
```
Comment (4144) [inline src/map/packets_struct.h:5820]: packet should start from 20140611

File: src/map/goldpc.c [hunk 6]
```diff
@@ -33,8 +34,110 @@
...
+	idb_put(goldpc->db, mode.id, mode_entry);
+
+	script->set_constant2(const_name, mode.id, false, false);
+
+	return true;
+}
+
+bool goldpc_read_db_validate(struct goldpc_mode *mode, const char *source)
+{
+	nullpo_retr(false, mode);
```
Comment (4144) [inline src/map/goldpc.c:116]: missing nullpo for source

**Review by 4144 — CHANGES_REQUESTED** [review]

File: src/map/packets_struct.h (hunk 5 above)
Comment (guilherme-gm) [reply src/map/packets_struct.h]: I kept this one for 20140430 as it is CZ packet

**Comment by guilherme-gm** [comment]
@4144 Thanks for the review, I added new "fixup" commits with the changes and will rebase them once everything is ok.

**Review by Asheraf — APPROVED** [review]

**Review by 4144 — APPROVED** [review]

**Comment by guilherme-gm** [comment]
rebased squashing the fixup commits into the ones that had issues

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**